samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java

```diff
@@ -0,0 +1,284 @@
+                    System.out.println("Enter the message you want to publish to topic %s and press Enter. " +
+                            "Type 'exit' or 'quit' to exit this program.");
+                    input = scanner.nextLine();
+                }
+
+                if("exit".equals(input) || "quit".equals(input)) {
+                    System.out.println("Terminating...");
+                    break;
+                }
+
+                if("publish".equals(mode) || "both".equals(mode)) {
+                    System.out.println("Publishing message!");
+                    final CompletableFuture<Integer> publishResult = connection.publish(new MqttMessage(topic,
+                            input.getBytes(StandardCharsets.UTF_8)), QualityOfService.AT_LEAST_ONCE, false);
+                    Integer result = publishResult.get();
+                    System.out.println("Publish result code: " + result);
+                }
+            }
+        } catch (CrtRuntimeException | InterruptedException | ExecutionException ex) {
+            System.out.println("Exception encountered: " + ex.toString());
+            ex.printStackTrace();
+        }
+        finally {
+            if(connection != null && !connection.isNull()) {
+                CompletableFuture<Void> disconnected = connection.disconnect();
+                try {
+                    disconnected.get();
+                } catch (ExecutionException | InterruptedException e) {
+                    throw new RuntimeException("Problem disconnecting the MQTT connection cleanly: " + e.getMessage());
+                }
+            }
+        }
+        System.out.println("Complete!");
+    }
+
+    private static MqttClientConnection getClientFromDiscovery(final DiscoveryClient discoveryClient,
+                                                               final ClientBootstrap bootstrap) throws ExecutionException, InterruptedException {
+        final CompletableFuture<DiscoverResponse> futureResponse = discoveryClient.discover(thingName);
+        final DiscoverResponse response = futureResponse.get();
+        if(response.getGGGroups() != null) {
+            final Optional<GGGroup> groupOpt = response.getGGGroups().stream().findFirst();
+            if(groupOpt.isPresent()) {
+                final GGGroup group = groupOpt.get();
+                final GGCore core = group.getCores().stream().findFirst().get();
+                final String dnsOrIp = core.getConnectivity().get(0).getHostAddress();
+                final Integer port = core.getConnectivity().get(0).getPortNumber();
+                System.out.println(String.format("Connecting to group ID %s, with thing arn %s, using endpoint %s:%d",
+                        group.getGGGroupId(), core.getThingArn(), dnsOrIp, port));
+
+                final AwsIotMqttConnectionBuilder connectionBuilder = AwsIotMqttConnectionBuilder.newMtlsBuilderFromPath(certPath, keyPath)
```

Comment:
This should be try-blocked and the returned connection be assigned within a try-block I think.